### PR TITLE
Limit number of external services per org/user

### PR DESCRIPTION
**Pull request by miveronese**
This PR adds, to the backend, on the Cloud mode, a limit to external services, allowing a max of 1 Github service + 1 Gitlab service per org/user.

As a result:
 - the API cannot be used to add more than 1 external_service of a given type  
 - the API cannot be used to add external services that are not currently supported (only Github and Gitlab are currently supported)

#### Unit tests
 New unit tests have been added to cover the new checks. You can run them with `go test ./cmd/frontend/graphqlbackend/. -v -run TestAddExternalService`

#### Manual tests
Manual tests were done (see video) using the /api/console with the following mutation. 

```
mutation AddExternalService($input: AddExternalServiceInput!) {
	addExternalService(input: $input) {
		kind,
    displayName,
    config,
  }
}
```
Query Variables:
```
{
  "input": {
    "kind": "Test",
    "displayName": "Test",
    "namespace": "Test",
    "config": "{\n  \"url\": \"https://test.com\",\n  \"token\": \"testest\",\n  \"projectQuery\": [\n    \"projects?membership=true&archived=no\"\n  ]\n}"
  }
}
```

https://user-images.githubusercontent.com/1552863/162556880-4cd634c5-30b8-4400-9e9b-554cf4f57aa9.mov

### CODE REVIEW

File: cmd/frontend/graphqlbackend/external_services.go [hunk 1]
```diff
@@ -86,6 +87,22 @@ func (r *schemaResolver) AddExternalService(ctx context.Context, args *addExtern
 		return nil, err
 	}
 
+	// TODO: OPEN QUESTIONS
```
Comment (miveronese) [inline cmd/frontend/graphqlbackend/external_services.go:90]: Team (@sourcegraph/cloud-saas) , the following are the questions that I need to clarify...

Comment (unknwon) [reply cmd/frontend/graphqlbackend/external_services.go:90]: 1. Yes
2. Yes
3. Don't we do? https://sourcegraph.com/site-admin

I think the limit only applies to user and org code host connections (either have namespaceOrgID or namespaceUserID `!= nil`)

Comment (kopancek) [reply cmd/frontend/graphqlbackend/external_services.go:90]: 1. yes. It's either owned by org, or user or neither. Cannot be owned by both a user and an org.
2. Line 56 is a branch of code for namespaced external services - owned by an org or a user. Line 85 is for the other branch - no namespace provided - so site-level external services. Only site-admins can add those.
3. We do have site admin even in Cloud. Maybe you're just not a site admin at the moment?

Comment (miveronese) [reply cmd/frontend/graphqlbackend/external_services.go:90]: Thank you @unknwon  and @kopancek! I wasn't familiar with the site-level external services (option, granted to site admins, to add a service without providing a namespace). My understanding is the same as @unknwon 's (correct me, please, if I am mistaken): in the case of site-level external services  we don't need to apply a limit...

File: internal/extsvc/types.go [hunk 2]
```diff
@@ -581,3 +581,5 @@ func UniqueCodeHostIdentifier(kind, config string) (string, error) {
 
 	return NormalizeBaseURL(u).String(), nil
 }
+
+const LimitPerKind = 1
```
Comment (unknwon) [inline internal/extsvc/types.go:585]: Do we really need a constant here in this package? Its purpose is cloud-specific and the naming doesn't imply that, as well as it is very far from where (and the only places) it is actually used.

My suggestion would be to move this constant into the `CheckExternalServicesQuota` function.

File: cmd/frontend/graphqlbackend/external_services_test.go [hunk 3]
```diff
@@ -281,6 +282,256 @@ func TestAddExternalService(t *testing.T) {
...
+			}
+
+			externalServices := database.NewMockExternalServiceStore()
+			externalServices.ListFunc.SetDefaultReturn(svcs, nil)
+			db.ExternalServicesFunc.SetDefaultReturn(externalServices)
+
+			result, err := newSchemaResolver(db).AddExternalService(ctx, &addExternalServiceArgs{
+				Input: addExternalServiceInput{
+					Namespace:   &marshaledUserID,
+					Kind:        "GITLAB",
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/external_services_test.go:521]: I think we have a constant for this:

```suggestion
					Kind:        extsvc.KindGitLab,
```

Same for other occurrences.

File: cmd/frontend/backend/external_services.go [hunk 4]
```diff
@@ -43,3 +48,72 @@ func CheckOrgExternalServices(ctx context.Context, db database.DB, orgID int32)
...
+// reached. Max of 2 services - one for GitHub, one for GitLab - can be added per org or user
+func CheckExternalServicesQuota(ctx context.Context, db database.DB, kind string, orgID, userID int32) error {
+	services, err := servicesCountPerKind(ctx, db, orgID, userID)
+	if err != nil {
+		return err
+	}
+
+	if services[extsvc.KindGitHub] >= extsvc.LimitPerKind && services[extsvc.KindGitLab] >= extsvc.LimitPerKind {
+		return ErrExternalServicesQuotaReached
+	}
```
Comment (unknwon) [inline cmd/frontend/backend/external_services.go:62]: What is the rationale or benefit of having this extra check? i.e. wouldn't following two checks serve the same purpose?

File: cmd/frontend/backend/external_services.go [hunk 5]
```diff
@@ -43,3 +48,72 @@ func CheckOrgExternalServices(ctx context.Context, db database.DB, orgID int32)
...
+			return ErrExternalServiceLimitPerKindReached
+		}
+		return nil
+	}
+
+	return nil
+}
+
+// servicesMap returns a map with the total count for each type of service
+func servicesCountPerKind(ctx context.Context, db database.DB, orgID, userID int32) (map[string]int, error) {
```
Comment (unknwon) [inline cmd/frontend/backend/external_services.go:82]:
```suggestion
// servicesCountPerKind returns a map with the total count for each type of service
func servicesCountPerKind(ctx context.Context, db database.DB, orgID, userID int32) (map[string]int, error) {
```

**Review by unknwon — COMMENTED** [review]
I think the code would work, but I can't really give my approval until we improve a bit more on the "style" 😁

File: cmd/frontend/backend/external_services.go (hunk 4 above)
Comment (miveronese) [reply cmd/frontend/backend/external_services.go]: As we allow a max of 2 services, one per type, two cases can happen:

a- user reaches the limit for one kind but not for another.
b- user reaches the limit for both kinds. 

Line 60-62 covers B, while the following cover A. 
So, the goal here is making it clearer which limit was reached: total (all services) or per kind (just one service).

Comment (unknwon) [reply cmd/frontend/backend/external_services.go]: On a second look, this function is specifically asking for a `Kind` to check with, just felt doing a bit too much but a nit anyway.

**Review by unknwon — APPROVED** [review]

File: cmd/frontend/backend/external_services.go (hunk 4 above)
Comment (miveronese) [reply cmd/frontend/backend/external_services.go]: yeah, I see your point, @unknwon!  Will update this part before merging.

**Pull request merged**